paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java

```diff
@@ -203,7 +231,35 @@ public AdjustedProjection adjustProjection(@Nullable int[][] projection) {
             if (fieldSequences.isEmpty()) {
                 return new AdjustedProjection(projection, null);
             }
-            // TODO implement this, just keep required fields and adjust fieldSequences too.
+
+            if (projection != null) {
+                int[] projIndexs = Projection.of(projection).toTopLevelIndexes();
```

Comment:
```
            if (projection == null) {
                return new AdjustedProjection(null, null);
            }
            LinkedHashSet<Integer> extraFields = new LinkedHashSet<>();
            int[] topProjects = Projection.of(projection).toTopLevelIndexes();
            for (int index : topProjects) {
                SequenceGenerator generator = fieldSequences.get(index);
                if (generator != null) {
                    extraFields.add(generator.index());
                }
            }

            int[] allProjects = Stream.concat(
                    Arrays.stream(topProjects).boxed(),
                    extraFields.stream()).mapToInt(Integer::intValue).toArray();

            int[][] pushdown = Projection.of(allProjects).toNestedIndexes();
            int[][] outer = Projection.of(IntStream.range(0, topProjects.length).toArray()).toNestedIndexes();
            return new AdjustedProjection(pushdown, outer);
```

Reply:
There may be some sequence fields in `topProjects`, we don't need to add these fields to `extraFields`.